packages/web3-ccip-read/src/index.js

```diff
@@ -0,0 +1,210 @@
+var normalizeResponse = function (errorObject, result) {
+    const defaultResponse = {
+        data: ''
+    };
+
+    if (!errorObject && !result) {
+        return defaultResponse;
+    }
+
+    if (typeof errorObject === "string" && hasCcipReadFunctionSelector(errorObject)) {
+        return {
+            data: errorObject
+        };
+    }
+
+    if (typeof result === "string" && hasCcipReadFunctionSelector(result)) {
+        return {
+            data: result
+        };
+    }
+
+    if (
+        typeof errorObject === 'object' &&
+        hasCcipReadFunctionSelector(errorObject && errorObject.data)
+    ) {
+        return {
+            data: errorObject.data
+        };
+    }
+
+    return defaultResponse;
+};
+
+/**
+ * Determine if revert is a CCIP-Read error
+ *
+ * @method isOffChainLookup
+ *
+ * @param {Error} err
+ * @param {Object} result
+ *
+ * @return {Boolean} true if reversion was a CCIP-Read error
+ */
+var isOffChainLookup = function (err, result) {
+    const normalizedResponse = normalizeResponse(err, result);
+    return !!normalizedResponse.data;
+};
+
+/**
+ * Should be used to encode list of params
+ *
+ * @method encodeParameters
+ *
+ * @param {Array<String|Object>} types
+ * @param {Array<any>} params
+ *
+ * @return {String} encoded list of params
+ */
+var ccipReadCall = async function (errorObject, result, payload, send, options) {
+    if (send.ccipReadCalls) {
+        send.ccipReadCalls++;
+    } else {
+        send.ccipReadCalls = 1;
+    }
+    if (send.ccipReadCalls > MAX_REDIRECT_COUNT) {
+        throw new Error('Too many CCIP-read redirects');
+    }
+
+    const normalizedResponse = normalizeResponse(errorObject, result);
+    if (!normalizedResponse.data) {
+        throw new Error('ccipReadCall called for a non-CCIP-read compliant error');
+    }
+
+    const [sender, urls, callData, callbackFunction, extraData] = Object.values(
+        abi.decodeParameters(OFFCHAIN_LOOKUP_PARAMETER_TYPES, `${normalizedResponse.data.substring(10)}`)
+    );
+
+    if (
+        (sender && sender.toLowerCase()) !==
+        (payload && payload.params && payload.params[0] && payload.params[0].to && payload.params[0].to.toLowerCase())
+    ) {
+        throw new Error('CCIP-read error: sender does match contract address');
+    }
+
+
+    let finalUrls;
+    if(options.ccipReadGatewayUrls.length) {
+        finalUrls = options.ccipReadGatewayUrls;
+    } else {
+        finalUrls = urls;
+    }
+
+    if(!finalUrls.length) {
+        throw new Error('No gateway urls provided');
+    }
+
+    let gatewayResult;
+    if(options.ccipReadGatewayCallback) {
+        try{
+            gatewayResult = await options.ccipReadGatewayCallback(finalUrls, sender, callData);
+        } catch(e) {
+            console.error('ccipReadGatewayCallback error.');
+            throw e;
+        }
+    } else {
+        const result = await callGateway(finalUrls, sender, callData);
```

Comment:
You should probably export `callGateway` so that implementers of `ccipReadGatewayCallback` can use it.